stdlib/seq.ml

```diff
@@ -50,16 +52,24 @@ let rec filter f seq () = match seq() with
       then Cons (x, filter f next)
       else filter f next ()
 
-let rec flat_map f seq () = match seq () with
+let rec flatten seq () = match seq () with
   | Nil -> Nil
   | Cons (x, next) ->
-    flat_map_app f (f x) next ()
+     append_flatten x next ()
+(* [append_flatten seq tail] is [append seq (flatten tail)] *)
+and append_flatten seq tail () = match seq () with
+  | Nil -> flatten tail ()
+  | Cons (x, next) -> Cons (x, append_flatten next tail)
 
-(* this is [append seq (flat_map f tail)] *)
-and flat_map_app f seq tail () = match seq () with
+let rec flat_map f seq () = match seq () with
+  | Nil -> Nil
+  | Cons (x, next) ->
+    append_flat_map (f x) f next ()
+(* [append_flat_map seq f tail] is [append seq (flat_map f tail)] *)
+and append_flat_map seq f tail () = match seq () with
   | Nil -> flat_map f tail ()
   | Cons (x, next) ->
-    Cons (x, flat_map_app f next tail)
+    Cons (x, append_flat_map next f tail)
```

Comment:
What's the benefit of auxiliary mutually recursive functions for `flatten` and `flat_map` over simpler implementations below? 

```
  let rec concat (s1 : 'a t) (s2 : 'a t) : 'a t =
   fun () ->
    match s1 () with Nil -> s2 () | Cons (v, rest) -> Cons (v, concat rest s2)

  let rec flatten (ss : 'a t t) : 'a t =
   fun () ->
    match ss () with
    | Nil -> Nil
    | Cons (v, rest) -> (concat v (flatten rest)) ()

  let rec flat_map (f : 'a -> 'b t) (s : 'a t) : 'b t =
   fun () ->
    match s () with
    | Nil -> Nil
    | Cons (v, rest) -> concat (f v) (flat_map f rest) ()
```

`concat` would be `Seq.append`.

Reply:
I don't have a very convincing answer to this question, so I think I will just rewrite the code as you suggest.

Somewhat-explanatory remarks:
- The code of flat_map was written in this style originally by @c-cube, so it felt natural to reuse the structure
- When writing standard-library code, it is somewhat common to write in a fairly low-level style with some parts manually inlined (but here I could not convince myself that it might have any performance benefit).
- In strict versions of the same functions (for lists typically), the inlining has the benefit that it exposes an opportunity for TRMC transformation (see #9760). Here this aspect is irrelevant as recursive calls return a function closure immediately, so the call-stack usage remains constant.

Reply:
Thanks for remarks! It's a pity that trmc PR didn't get merged

Reply:
I changed the implementation as suggested. Thanks!